detox/android/detox/src/main/java/com/wix/detox/espresso/scroll/SwipeHelper.kt

```diff
@@ -0,0 +1,103 @@
+package com.wix.detox.espresso.scroll
+
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.action.*
+import com.wix.detox.espresso.common.annot.*
+import kotlin.math.min
+import kotlin.math.max
+import kotlin.math.sign
+
+typealias Point = Pair<Double, Double>
+
+object SwipeHelper {
+    fun swipeFastInDirection(direction: Int): ViewAction? {
+        return when (direction) {
+            MOTION_DIR_LEFT -> ViewActions.swipeLeft()
+            MOTION_DIR_RIGHT -> ViewActions.swipeRight()
+            MOTION_DIR_UP -> ViewActions.swipeUp()
+            MOTION_DIR_DOWN -> ViewActions.swipeDown()
+            else -> throw RuntimeException("Unsupported swipe direction: $direction")
+        }
+    }
+
+    fun swipeCustomInDirection(direction: Int, fast: Boolean, offset: Double, startPositionX: Double, startPositionY: Double): ViewAction? {
+        val swiper = if (fast) Swipe.FAST else Swipe.SLOW
+        val (from, to) = calculateSwipe(direction, offset, startPositionX, startPositionY)
+        val start = translate(from)
+        val end = translate(to)
+        val swipeAction = GeneralSwipeAction(swiper, start, end, Press.FINGER)
+        return ViewActions.actionWithAssertions(swipeAction)
+    }
+
+    fun calculateSwipe(
+            direction: Int,
+            unsafeOffset: Double = Double.NaN,
+            unsafeStartX: Double = Double.NaN,
+            unsafeStartY: Double = Double.NaN
+    ): Pair<Point, Point> {
+        val isHorizontal = direction == MOTION_DIR_LEFT || direction == MOTION_DIR_RIGHT
+        val isVertical = direction == MOTION_DIR_UP || direction == MOTION_DIR_DOWN
+        val isDescending = direction == MOTION_DIR_UP || direction == MOTION_DIR_LEFT
+
+        if (!isHorizontal && !isVertical) {
+            throw RuntimeException("Unsupported swipe direction: $direction")
+        }
+
+        var offset = safeProportion(unsafeOffset, 0.5 + EDGE_FUZZ_FACTOR);
+        if (isDescending) {
+            offset *= -1
+        }
+
+        val startPrimary = safeProportion(
+                if (isHorizontal) unsafeStartX else unsafeStartY,
+                0.5 - sign(offset) * EDGE_FUZZ_FACTOR
+        )
+
+        val startSecondary = safeProportion(
```

Comment:
Perhaps rename to `startPrimaryAxis`, etc.? it got me confused, thinking there might be two competing start points, or an intermediate point along the axis 😬

Reply:
maybe even swipeAxis and dormantAxis? 😅